### PR TITLE
Add a state check capability for auxkernel/uo evaluations

### PR DESCRIPTION
**Pull request by YaqiWang**
Refer to  #22882.

We possibly can fix all failing tests and then turn this parameter on at least for framework tests. The failing tests I saw on my mac:

```
userobjects/interface_user_object.interface_UO_QP_test ....................................... FAILED (ERRMSG)
postprocessors/old_vpp_value.old_no_history .................................................. FAILED (ERRMSG)
userobjects/pre_aux_based_on_exec_flag.test_preaux_based_on_exec_flag_final .................. FAILED (ERRMSG)
vectorpostprocessors/late_declaration_vector_postprocessor.test .............................. FAILED (ERRMSG)
userobjects/force_aux_ordering.test_force/postaux ............................................ FAILED (ERRMSG)
userobjects/interface_user_object.interface_Qp_rate_incremement_test ......................... FAILED (ERRMSG)
materials/material_dependency.dont_reinit_mat_for_aux ........................................ FAILED (ERRMSG)
userobjects/element_subdomain_modifier.group/stateful_property ............................... FAILED (ERRMSG)
reporters/base.base ............................................................. [min_cpus=2] FAILED (ERRMSG)
materials/stateful_prop.stateful_copy ........................................................ FAILED (ERRMSG)
interfaces/random.material_serial ............................................................ FAILED (ERRMSG)
postprocessors/change_over_fixed_point.change_over_fixed_point/change_with_respect_to_previous FAILED (CSVDIFF)
mesh/face_info.face_info/quad ............................................................... FAILED (CSVDIFF)
postprocessors/random_pps.random_pps/full_sim ............................................... FAILED (CSVDIFF)
userobjects/force_aux_ordering.test_force/preaux ............................................ FAILED (CSVDIFF)
vectorpostprocessors/dynamic_point_sampler.dynamic_point_sampler ............................ FAILED (CSVDIFF)
outputs/reporters.csv/root .................................................................. FAILED (CSVDIFF)
mesh/boundary_subdomain_list.boundary_subdomain_connections ................................. FAILED (CSVDIFF)
executioners/adapt_and_modify.test .......................................................... FAILED (EXODIFF)
vectorpostprocessors/least_squares_fit_history.least_squares_history ........................ FAILED (CSVDIFF)
vectorpostprocessors/point_value_sampler_history.test ....................................... FAILED (CSVDIFF)
mesh/face_info.face_info/tri ................................................................ FAILED (CSVDIFF)
auxkernels/pp_depend.pp_depend_indirect_wrong ............................................... FAILED (CSVDIFF)
mesh/face_info.face_info/two_region_quads ................................................... FAILED (CSVDIFF)
executioners/arbitrary_execute_flag.test .................................................... FAILED (CSVDIFF)
dirackernels/reporter_point_source.reporterPointSource/2dTransient .......................... FAILED (CSVDIFF)
userobjects/setup_interface_count.setup_interface_count/GeneralUserObject ...... [min_cpus=2] FAILED (CSVDIFF)
userobjects/setup_interface_count.setup_interface_count/ElementUserObject ................... FAILED (CSVDIFF)
userobjects/setup_interface_count.setup_interface_count/SideUserObject ...................... FAILED (CSVDIFF)
userobjects/setup_interface_count.setup_interface_count/InternalSideUserObject .............. FAILED (CSVDIFF)
userobjects/setup_interface_count.setup_interface_count/NodalSideUserObject ................. FAILED (CSVDIFF)
interfaces/random.test_uo ................................................................... FAILED (EXODIFF)
```

Some of them might be real issues.

### CODE REVIEW

**Comment by friedmud** [comment]
Can you tag the issue you are working on?  I would like to see some more discussion of what this is for.  Thanks!

**Comment by zachmprince** [comment]
I think it would be a little cleaner with this at the bottom of `FEProbomeBase::execute`:

```cpp
  if (_uo_aux_state_check && !_checking_uo_aux_state)
  {
    std::unique_ptr<NumericVector<Number>> x = _aux->currentSolution->clone();
    _checking_uo_aux_state = true;
    execute(exec_type);
    _checking_uo_aux_state = false;
    *x -= *_aux->currentSolution();
    if (x->l2_norm() > 1e-8)
      mooseError("Aux variables, user objects appear to have state on ", exec_type);
  }
```

We should also check PPs and VPPs somehow since we could have circular dependencies without auxkernels.

**Comment by YaqiWang** [comment]
> I think it would be a little cleaner with this at the bottom of `FEProbomeBase::execute`:
> 
> ```c++
>   if (_uo_aux_state_check && !_checking_uo_aux_state)
>   {
>     std::unique_ptr<NumericVector<Number>> x = _aux->currentSolution->clone();
>     _checking_uo_aux_state = true;
>     execute(exec_type);
>     _checking_uo_aux_state = false;
>     *x -= *_aux->currentSolution();
>     if (x->l2_norm() > 1e-8)
>       mooseError("Aux variables, user objects appear to have state on ", exec_type);
>   }
> ```
> 
> We should also check PPs and VPPs somehow since we could have circular dependencies without auxkernels.

I like this suggestion and will borrow it ;-). To check PPs and VPPs, we possibly will need to check reporter data. Let me spent a little more time on whether we can just check the two quickly.

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -3954,6 +3961,64 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
 
   // Return the current flag to None
   setCurrentExecuteOnFlag(EXEC_NONE);
+
+  if (_uo_aux_state_check && !_checking_uo_aux_state)
+  {
+    // we will only check aux variables and postprocessors
+    // checking more reporter data can be added in the future if needed
+    std::unique_ptr<NumericVector<Number>> x = _aux->currentSolution()->clone();
```
Comment (zachmprince) [inline framework/src/problems/FEProblemBase.C:3969]: Looks like you are allocating and deallocating this vector every time this is called. We could add a vector to aux system to avoid this, but it create another dreaded member in FEProblemBase

File: framework/src/reporters/ReporterData.C [hunk 2]
```diff
@@ -52,6 +52,53 @@ ReporterData::getPostprocessorNames() const
...
+{
+  DenseVector<Real> all_values;
+
+  std::vector<Real> & output = all_values.get_values();
+
+  for (const auto & name_context_pair : _context_ptrs)
+  {
+    if (name_context_pair.first.isPostprocessor())
+      output.push_back(
+          getReporterValue<PostprocessorValue>(name_context_pair.first.getCombinedName()));
```
Comment (zachmprince) [inline framework/src/reporters/ReporterData.C:66]: This should capture PPs and scalar reporter values.
```suggestion
          getReporterValue<Real>(name_context_pair.first.getCombinedName()));
```

File: framework/src/reporters/ReporterData.C [hunk 3]
```diff
@@ -52,6 +52,53 @@ ReporterData::getPostprocessorNames() const
...
+
+    if (name_context_pair.first.isVectorPostprocessor())
+    {
+      auto & vec =
+          getReporterValue<VectorPostprocessorValue>(name_context_pair.first.getCombinedName());
```
Comment (zachmprince) [inline framework/src/reporters/ReporterData.C:71]:
```suggestion
      const auto & vec =
          getReporterValue<std::vector<Real>>(name_context_pair.first.getCombinedName());
```

File: framework/src/problems/FEProblemBase.C [hunk 4]
```diff
@@ -3954,6 +3961,64 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+      for (const auto & pair : ordered_map)
+        oss << "  {" << pair.second << ", " << pair.first << "},\n";
+
+      mooseError("Aux kernels, user objects appear to have states for aux variables on ",
+                 exec_type,
+                 ".\nVariable error norms in descending order:\n",
+                 oss.str());
+    }
+
+    pp_values -= getReporterData().getAllPostprocessorValues();
```
Comment (zachmprince) [inline framework/src/problems/FEProblemBase.C:3996]: Should probably check to size match before doing this operation.

File: framework/src/reporters/ReporterData.C [hunk 5]
```diff
@@ -52,6 +52,53 @@ ReporterData::getPostprocessorNames() const
...
+    if (name_context_pair.first.isPostprocessor())
+      output.push_back(name_context_pair.first.getCombinedName());
+
+    if (name_context_pair.first.isVectorPostprocessor())
+    {
+      auto pname = name_context_pair.first.getCombinedName();
+      auto & vec = getReporterValue<VectorPostprocessorValue>(pname);
+      for (unsigned int i = 0; i < vec.size(); ++i)
+        output.push_back(pname + "/" + std::to_string(i));
+    }
```
Comment (zachmprince) [inline framework/src/reporters/ReporterData.C:96]:
```suggestion
    const ReporterName & rname = name_context_pair.first;
    
    if (hasReporterValue<Real>(rname))
      output.push_back(rname);

    if (hasReporterValue<std::vector<Real>>(rname))
    {
      const auto & vec = getReporterValue<std::vector<Real>>(rname);
      for (unsigned int i = 0; i < vec.size(); ++i)
        output.push_back(std::to_string(rname) + "_" + std::to_string(i));
    }
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
Very nice addition! One more suggestion would be to add a test that triggers the errors. We should also document the parameter somewhere.

File: framework/src/reporters/ReporterData.C (hunk 2 above)
Comment (YaqiWang) [reply framework/src/reporters/ReporterData.C:66]: The above check `if (name_context_pair.first.isPostprocessor())` will be only for postprocessor, right?

File: framework/src/problems/FEProblemBase.C (hunk 4 above)
Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C:3996]: I will add an assert. Supposedly after the second call of `execute()` at the same execute, the size should not change.

File: framework/src/reporters/ReporterData.C (hunk 5 above)
Comment (YaqiWang) [reply framework/src/reporters/ReporterData.C:96]: I see what you want. I will need to change both functions and possibly rename the function a bit to reflect this change.

Comment (YaqiWang) [reply framework/src/reporters/ReporterData.C:96]: `/` makes the print nicer than `_` because it clearly separate the vector name and the index, otherwise the index looks like part of the name.

File: framework/src/problems/FEProblemBase.C (hunk 4 above)
Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C:3996]: I changed my mind, I will just add a check because this is not on performance critical code path and can make the code more protective.

**Comment by YaqiWang** [comment]
Because I am going to add a test as suggested by @zachmprince , I will drop the default value change so that all tests can pass. Developers can easily reproduce the errors by `./run_tests -j12 --cli-args Problem/check_uo_aux_state=true`. Fixing all failing tests can be in a separate PR in the future.

**Comment by YaqiWang** [comment]
@zachmprince your comments have been addressed or answered. Thanks for the review!

**Comment by YaqiWang** [comment]
The failure seems irrelevant. I think this PR is ready for review again. If this can be merged, I can make Griffin this check free quickly.

**Comment by YaqiWang** [comment]
@zachmprince one thing I forgot mentioning to you is that this flag fails STM, which I have not looked into, although it seems  not supposed to.

File: framework/src/problems/FEProblemBase.C [hunk 6]
```diff
@@ -3954,6 +3961,72 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+      auto pp_names = getReporterData().getAllRealReporterFullNames();
+      std::multimap<Real, std::string, std::greater<Real>> ordered_map;
+      for (const auto i : index_range(pp_names))
+        ordered_map.insert(std::pair<Real, std::string>(std::abs(pp_values(i)), pp_names[i]));
+
+      std::ostringstream oss;
+      unsigned int counter = 0;
+      for (const auto & pair : ordered_map)
+      {
+        if (counter == 11 || pair.first < 1e-8)
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:4016]: This counter value seems so random

File: framework/src/problems/FEProblemBase.C [hunk 7]
```diff
@@ -3954,6 +3961,72 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+                 oss.str());
+    }
+
+    DenseVector<Real> new_pp_values = getReporterData().getAllRealReporterValues();
+    if (pp_values.size() != new_pp_values.size())
+      mooseError("Second execution for uo/aux state check should not change the number of "
+                 "real reporter values");
+
+    pp_values -= new_pp_values;
+    if (pp_values.l2_norm() > 1e-8)
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:4005]: I think if you want this to work in parallel as well where there may be different floating point operation orderings we should do more of a relative check instead of absolute

File: framework/src/problems/FEProblemBase.C [hunk 8]
```diff
@@ -3954,6 +3961,72 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+    {
+      auto pp_names = getReporterData().getAllRealReporterFullNames();
+      std::multimap<Real, std::string, std::greater<Real>> ordered_map;
+      for (const auto i : index_range(pp_names))
+        ordered_map.insert(std::pair<Real, std::string>(std::abs(pp_values(i)), pp_names[i]));
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:4010]:
```suggestion
        ordered_map.emplace(std::abs(pp_values(i)), pp_names[i]);
```

File: test/tests/auxkernels/pp_depend/tests [hunk 9]
```diff
@@ -9,15 +9,16 @@
...
 
   [./pp_depend_indirect_wrong]
-    type = 'CSVDiff'
+    type = 'RunException'
     input = 'pp_depend_indirect_wrong.i'
-    csvdiff = 'pp_depend_indirect_wrong_out.csv'
-    requirement = "Execution of a FunctionValuePostprocessor that depends on a Function which depends on a second PostProcessor is construction order dependent."
+    cli_args = 'Problem/check_uo_aux_state=true'
+    expect_err = 'Aux kernels, user objects appear to have states for real reporter values on TIMESTEP_END'
+    requirement = "The system shall be able to detect wrong implicit dependencies of a FunctionValuePostprocessor that depends on a Function which depends on a second PostProcessor."
```
Comment (lindsayad) [inline test/tests/auxkernels/pp_depend/tests:16]: requirements should not have implementation details like class names

File: test/tests/auxkernels/pp_depend/tests [hunk 10]
```diff
@@ -9,15 +9,16 @@
...
   [../]
   [./pp_depend_indirect_correct]
     type = 'CSVDiff'
     input = 'pp_depend_indirect_correct.i'
     csvdiff = 'pp_depend_indirect_correct_out.csv'
-    requirement = "Implicit dependencies of FunctionValuePostprocessors may be declared to resolve execution order independently of construction order."
+    requirement = "Explicit dependencies of FunctionValuePostprocessors may be declared to resolve execution order independently of construction order."
```
Comment (lindsayad) [inline test/tests/auxkernels/pp_depend/tests:22]: same

File: test/tests/userobjects/pre_aux_based_on_exec_flag/tests [hunk 11]
```diff
@@ -1,12 +1,20 @@
...
-  issues = '#18189'
-  design = '/GeneralUserObject.md'
+  [state_detection]
+    type = 'RunException'
+    input = 'pre_post_aux_test.i'
+    cli_args = 'Problem/check_uo_aux_state=true AuxKernels/inactive=NormalizationAuxINITIAL'
+    expect_err = 'Aux kernels, user objects appear to have states for aux variables on FINAL'
+    issues = '#22802'
+    design = 'FEProblemBase.md'
+    requirement = 'The system shall provide means to detect unresolved dependency in evaluations of user objects and auxiliary kernels '
```
Comment (lindsayad) [inline test/tests/userobjects/pre_aux_based_on_exec_flag/tests:9]: This is a good requirement

**Review by lindsayad — DISMISSED** [review]

File: framework/src/problems/FEProblemBase.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:3969]: I agree with @YaqiWang that since this integrity check is default off, if a user turns it on then we should not care too much about performance.

File: framework/src/reporters/ReporterData.C [hunk 12]
```diff
@@ -52,6 +52,55 @@ ReporterData::getPostprocessorNames() const
...
+
+    if (hasReporterValue<Real>(rname))
+      output.push_back(getReporterValue<Real>(rname.getCombinedName()));
+
+    if (hasReporterValue<std::vector<Real>>(rname))
+    {
+      const auto & vec = getReporterValue<std::vector<Real>>(rname.getCombinedName());
+      for (const auto & v : vec)
+        output.push_back(v);
+    }
```
Comment (lindsayad) [inline framework/src/reporters/ReporterData.C:74]: Would be nice to include a vector postprocessor in your test in order to hit these lines but not critical

File: framework/src/reporters/ReporterData.C [hunk 13]
```diff
@@ -52,6 +52,55 @@ ReporterData::getPostprocessorNames() const
...
+    if (hasReporterValue<Real>(rname))
+      output.push_back(rname.getCombinedName());
+
+    if (hasReporterValue<std::vector<Real>>(rname))
+    {
+      auto pname = rname.getCombinedName();
+      const auto & vec = getReporterValue<std::vector<Real>>(pname);
+      for (unsigned int i = 0; i < vec.size(); ++i)
+        output.push_back(pname + "/" + std::to_string(i));
+    }
```
Comment (lindsayad) [inline framework/src/reporters/ReporterData.C:98]: Other lines missing coverage without a vector postprocessor. Not critical

File: framework/src/reporters/ReporterData.C (hunk 12 above)
Comment (YaqiWang) [reply framework/src/reporters/ReporterData.C:74]: Will just turn on this flag for a vector postprocessor test.

File: framework/src/problems/FEProblemBase.C (hunk 6 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:4016]: If we're going to error out anyway, might as well display all the problems

Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C:4016]: Agreed, plus it will make the code simpler.

**Comment by YaqiWang** [comment]
@lindsayad added one commit for addressing your comments.

**Comment by lindsayad** [comment]
How do we hit the vector specific code of `getAllRealReporterValues` but not the vector specific code of `getAllRealReporterFullNames`?

**Comment by lindsayad** [comment]
@zachmprince can you see whether your review has been addressed and resolve the conversations you started as you see fit?

**Comment by YaqiWang** [comment]
> How do we hit the vector specific code of `getAllRealReporterValues` but not the vector specific code of `getAllRealReporterFullNames`?

That is probably due to not getting into the error if statement where `getAllRealReporterFullNames` is called.

**Comment by lindsayad** [comment]
that makes sense

**Review by zachmprince — DISMISSED** [review]

File: framework/doc/content/source/problems/FEProblemBase.md [hunk 14]
```diff
@@ -147,10 +149,16 @@ $\psi_8$
 - MOOSE `Kernels` must provide each of the terms in square brackets (evaluated at $\vec{x}_{q}$ or $\vec x_{q_{face}}$ as necessary).
 
 
-!syntax description /Problem/FEProblem
-
 !syntax parameters /Problem/FEProblem
 
+!alert note
+When [!param](/Problem/FEProblem/check_uo_aux_state) is set to true, MOOSE will evaluate user objects
+(including all postprocessors and vector postprocessors) and auxiliary kernels on every execute flag except `linear`
```
Comment (lindsayad) [inline framework/doc/content/source/problems/FEProblemBase.md:156]: Why is `linear` treated specially? Is it because of the PJFNK check we have in libMesh? It seems like linear would be almost the most concerning if there are issues there

File: framework/src/problems/FEProblemBase.C [hunk 15]
```diff
@@ -3954,6 +3961,68 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+    _checking_uo_aux_state = false;
+
+    const Real check_tol = 1e-8;
+
+    const Real xnorm = x->l2_norm();
+    *x -= *_aux->currentSolution();
+    if (x->l2_norm() > check_tol * xnorm)
+    {
+      const auto & sys = _aux->system();
+      unsigned int n_vars = sys.n_vars();
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:3984]:
```suggestion
      const unsigned int n_vars = sys.n_vars();
```

File: framework/src/problems/FEProblemBase.C [hunk 16]
```diff
@@ -3954,6 +3961,68 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+      std::multimap<Real, std::string, std::greater<Real>> ordered_map;
+      for (unsigned int i = 0; i < n_vars; i++)
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:3986]:
```suggestion
      for (const auto i : make_range(n_vars))
```

File: framework/src/problems/FEProblemBase.C [hunk 17]
```diff
@@ -3954,6 +3961,68 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+      {
+        const Real vnorm = sys.calculate_norm(*x, i, DISCRETE_L2);
+        ordered_map.emplace(vnorm, sys.variable_name(i));
+      }
+
+      std::ostringstream oss;
+      for (const auto & pair : ordered_map)
+        oss << "  {" << pair.second << ", " << pair.first << "},\n";
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:3994]:
```suggestion
      for (const auto & [error_norm, var_name] : ordered_map)
        oss << "  {" << var_name << ", " << error_norm << "},\n";
```

File: framework/src/problems/FEProblemBase.C [hunk 18]
```diff
@@ -3954,6 +3961,68 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+
+      mooseError("Aux kernels, user objects appear to have states for aux variables on ",
+                 exec_type,
+                 ".\nVariable error norms in descending order:\n",
+                 oss.str());
+    }
+
+    DenseVector<Real> new_pp_values = getReporterData().getAllRealReporterValues();
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:4002]:
```suggestion
    const DenseVector<Real> new_pp_values = getReporterData().getAllRealReporterValues();
```

File: framework/src/problems/FEProblemBase.C [hunk 19]
```diff
@@ -3954,6 +3961,68 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+    if (pp_values.size() != new_pp_values.size())
+      mooseError("Second execution for uo/aux state check should not change the number of "
+                 "real reporter values");
+
+    const Real ppnorm = pp_values.l2_norm();
+    pp_values -= new_pp_values;
+    if (pp_values.l2_norm() > check_tol * ppnorm)
+    {
+      auto pp_names = getReporterData().getAllRealReporterFullNames();
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:4011]:
```suggestion
      const auto pp_names = getReporterData().getAllRealReporterFullNames();
```

File: framework/src/problems/FEProblemBase.C [hunk 20]
```diff
@@ -3954,6 +3961,68 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
...
+      std::multimap<Real, std::string, std::greater<Real>> ordered_map;
+      for (const auto i : index_range(pp_names))
+        ordered_map.emplace(std::abs(pp_values(i)), pp_names[i]);
+
+      std::ostringstream oss;
+      for (const auto & pair : ordered_map)
+        oss << "  {" << pair.second << ", " << pair.first << "},\n";
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:4018]:
```suggestion
      for (const auto & [error_norm, pp_name] : ordered_map)
        oss << "  {" << pp_name << ", " << error_norm << "},\n";
```

File: test/tests/userobjects/pre_aux_based_on_exec_flag/pre_post_aux_test.i [hunk 21]
```diff
@@ -60,7 +60,7 @@
     source_variable = u
     shift = -100.0
     normalization = 1.0
-    execute_on = 'INITIAL TIMESTEP_BEGIN TIMESTEP_END FINAL'
+    execute_on = 'INITIAL FINAL'
```
Comment (lindsayad) [inline test/tests/userobjects/pre_aux_based_on_exec_flag/pre_post_aux_test.i:63]: why these changes?

**Review by lindsayad — CHANGES_REQUESTED** [review]
Should be good to go after this

File: framework/doc/content/source/problems/FEProblemBase.md (hunk 14 above)
Comment (YaqiWang) [reply framework/doc/content/source/problems/FEProblemBase.md:156]: This just because `FEProblemBase::computeResidualTags` is not calling `FEProblemBase::execute(EXEC_LINEAR)` function. Same as `FEProblemBase::computeResidualAndJacobian`. I think for residual (or linear), we already had a check in place, similar like this.

File: framework/src/problems/FEProblemBase.C (hunk 16 above)
Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C:3986]: I think `n_vars` is in unsigned int, so the old way is fine. But I will make this change.

File: test/tests/userobjects/pre_aux_based_on_exec_flag/pre_post_aux_test.i (hunk 21 above)
Comment (YaqiWang) [reply test/tests/userobjects/pre_aux_based_on_exec_flag/pre_post_aux_test.i:63]: Possibly unrelated trivial change. This aux kernel only depends on a primal variable thus does not have dependency issue. I possibly just made this change when trying to output on final.

File: framework/doc/content/source/problems/FEProblemBase.md (hunk 14 above)
Comment (YaqiWang) [reply framework/doc/content/source/problems/FEProblemBase.md:156]: Typically I think residual dependency issue in residual evaluations is caused by uo/auxkernels but not by kernels (unless kernels coupling tagged residual vectors, which we should prevent), so the existing check can be removed if we can call `execute(EXEC_LINEAR)`. This can be done along with the other `setup` cleanup in the future.

Comment (lindsayad) [reply framework/doc/content/source/problems/FEProblemBase.md:156]: Sad that `computeResidualTags` doesn't call `execute`

File: framework/src/problems/FEProblemBase.C (hunk 16 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C]: The change is robust against any changes to the type of `n_vars` and it ensures no one does anything funny with the loop counter

**Review by lindsayad — APPROVED** [review]
